lib/dockerregistry/checks.go

```diff
@@ -0,0 +1,133 @@
+func MKRealImageRemovalCheck(gitRepoURL string) *ImageRemovalCheck {
+	masterSHAStr := os.Getenv("PULL_BASE_SHA")
+	pullRequestSHAStr := os.Getenv("PULL_PULL_SHA")
+	masterSHA := plumbing.NewHash(masterSHAStr)
+	pullRequestSHA := plumbing.NewHash(pullRequestSHAStr)
+	return &ImageRemovalCheck{
+		gitRepoURL,
+		masterSHA,
+		pullRequestSHA,
+	}
+}
+
+// Run executes ImageRemovalCheck on a set of promotion edges.
+// Returns an error if the pull request removes images from the
+// promoter manifests.
+func (check *ImageRemovalCheck) Run(
+	edges map[PromotionEdge]interface{},
+) error {
+	var err error
+	// Get the Git repo the current working directory
+	r, err := gogit.PlainOpen(check.GitRepoURL)
+	if err != nil {
+		return fmt.Errorf("Could not open the Git repo: %v", err)
+	}
+
+	// Check out the master branch of the current Git repo to get the
+	// master branch's state of the promoter manifests
+	w, err := r.Worktree()
+	if err != nil {
+		return fmt.Errorf("Could not create Git worktree: %v", err)
+	}
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.MasterSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the master branch of the Git"+
+			" repo: %v", err)
+	}
+
+	// Parse the manifests and generate the promotion edges
+	mfests, err := ParseThinManifestsFromDir(check.GitRepoURL)
+	if err != nil {
+		return fmt.Errorf("Could not parse manifests from the directory: %v",
+			err)
+	}
+	masterEdges, err := ToPromotionEdges(mfests)
+	if err != nil {
+		return fmt.Errorf("Could not generate promotion edges from promoter"+
+			" manifests: %v", err)
+	}
+
+	// Reset the working Git repo by checking out the pull request
+	err = w.Checkout(&gogit.CheckoutOptions{
+		Hash:  check.PullRequestSHA,
+		Force: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Could not checkout the pull request branch of the"+
+			" Git repo %v: %v",
+			check.GitRepoURL, err)
+	}
+
+	return check.CompareEdges(edges, masterEdges)
+}
+
+// CompareEdges is a function of the ImageRemovalCheck that handles
+// the comparison of the pull requests's set of promotion edges and
+// the master branch's set of promotion edges.
+func (check *ImageRemovalCheck) CompareEdges(
+	edgesPullRequest map[PromotionEdge]interface{},
+	edgesMaster map[PromotionEdge]interface{},
+) error {
+	// Generate a set of all destination images that appear in
+	// the pull request's set of promotion edges
+	destinationImages := make(map[ImageTag]interface{})
+	for edge := range edgesPullRequest {
+		destinationImages[edge.DstImageTag] = nil
+	}
+
+	// Check that every destination image in the master branch's set of
+	// promotion edges exists in the pull request's set of promotion edges
+	removedImages := make([]string, 0)
+	for edge := range edgesMaster {
+		_, found := destinationImages[edge.DstImageTag]
+		if !found {
+			removedImages = append(removedImages,
+				string(edge.DstImageTag.ImageName))
+		}
+	}
```

Comment:
This check alone is not enough. Consider the case where we want to promote

(master edges) sha256:000 -> ImageTagX
(PR edges) sha256:111 -> ImageTagX

Now even though the PR is trying to promote a *different* digest, it is not detected by this check. That is, we actually *did* remove an image, with digest sha256:000 but this check did not detect it because it only looks at the ImageTag info and not the digests.